File: src/main/scala/us/zoom/data/dfence/providers/snowflake/revoke/evaluator/GrantRevocationEvaluator.scala
package us.zoom.data.dfence.providers.snowflake.revoke.evaluator

import org.slf4j.LoggerFactory
import us.zoom.data.dfence.extensions.*
import us.zoom.data.dfence.providers.snowflake.grant.builder.SnowflakeObjectType
import us.zoom.data.dfence.providers.snowflake.matchers.PlaybookGrantMatcher
import us.zoom.data.dfence.providers.snowflake.models.{
  PlaybookGrant,
  PrivilegeGrantIndex,
  SnowflakeGrant,
  SnowflakeGrantModel
}
import us.zoom.data.dfence.providers.snowflake.models.opaque.{GrantPrivilege, ObjectType, ObjectTypeAlias}

import scala.collection.parallel.ParSet
import scala.util.{Failure, Success, Try}

class GrantRevocationEvaluator(index: PrivilegeGrantIndex):

  private val log = LoggerFactory.getLogger(getClass)

  def needsRevoke(grant: SnowflakeGrantModel): Boolean =
    Try {
      val snowflakeGrant = SnowflakeGrant
        .from(grant)
        .getOrThrow(err =>
          throw IllegalArgumentException(
            s"Unable to create snow grant from snowflake grant model $grant",
            err
          )
        )

      val matcher = PlaybookGrantMatcher(snowflakeGrant)

      val matches = getCandidates(
        grantObjectType = snowflakeGrant.snowflakeObjectType,
        grantPrivilege = snowflakeGrant.privilege
      ).exists(matcher.matchesPlaybookGrant)

      !matches
    } match
      case Success(needsRevoke) =>
        if needsRevoke then log.info("Snowflake grant {} invalid, needs to be revoked", grant)
        else log.info("Snowflake grant {} is valid", grant)
        needsRevoke
      case Failure(err)         =>
        log.error("Error occurred while checking revocation of grants on model: {}", grant, err)
        false

  private def getCandidates(
    grantObjectType: SnowflakeObjectType,
    grantPrivilege: GrantPrivilege
  ): ParSet[PlaybookGrant] =
    getObjectTypeAndObjectTypeAliasMatchedGrants(
      grantObjectType
    ) & getPrivilegeMatchedGrants(grantPrivilege)

  private def getObjectTypeAndObjectTypeAliasMatchedGrants(
    grantObjectType: SnowflakeObjectType
  ): ParSet[PlaybookGrant] =
    val baseGrants = index
      .objectTypeIndex
      .getOrElse(
        ObjectType(grantObjectType),
        ParSet.empty
      )

    val specificGrants = index
      .objectAliasIndex
      .getOrElse(ObjectTypeAlias(grantObjectType), ParSet.empty)
      .filter(_.objectType == grantObjectType)

    baseGrants ++ specificGrants

  private def getPrivilegeMatchedGrants(
    grantPrivilege: GrantPrivilege
  ): ParSet[PlaybookGrant] =
    index.privilegeIndex.getOrElse(grantPrivilege, ParSet.empty)

File: src/main/java/us/zoom/data/dfence/providers/snowflake/revoke/evaluator/GrantRevocationEvaluator.java
package us.zoom.data.dfence.providers.snowflake.revoke.evaluator;

import io.vavr.control.Try;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import lombok.AccessLevel;
import lombok.AllArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import us.zoom.data.dfence.exception.RbacDataError;
import us.zoom.data.dfence.policies.models.PolicyGrant;
import us.zoom.data.dfence.policies.models.PolicyGrantPrivilege;
import us.zoom.data.dfence.providers.snowflake.grant.builder.SnowflakeObjectType;
import us.zoom.data.dfence.providers.snowflake.models.SnowflakeGrantModel;
import us.zoom.data.dfence.providers.snowflake.revoke.factories.SnowflakeGrantFactory;
import us.zoom.data.dfence.providers.snowflake.revoke.matchers.SnowflakeGrantMatchers;
import us.zoom.data.dfence.providers.snowflake.revoke.models.PolicyGrantHashIndex;
import us.zoom.data.dfence.providers.snowflake.revoke.models.SnowflakeGrant;

@Slf4j
@AllArgsConstructor(access = AccessLevel.PUBLIC)
public class GrantRevocationEvaluator {
  private final PolicyGrantHashIndex index;

  /**
   * Determines if a grant should be revoked. Returns true if the grant doesn't match any playbook
   * grant.
   */
  public boolean needsRevoke(SnowflakeGrantModel grant) {
    return Try.of(
            () -> {
              SnowflakeGrant snowflakeGrant = SnowflakeGrantFactory.createFrom(grant);
              return getCandidates(snowflakeGrant.snowflakeObjectType(), snowflakeGrant.privilege())
                  .stream()
                  .noneMatch(
                      policyGrant ->
                          SnowflakeGrantMatchers.matchesSnowflakeGrant()
                              .test(policyGrant, snowflakeGrant));
            })
        .getOrElseThrow(
            e -> {
              log.error("Failed to evaluate grant revocation for grant: {}", grant, e);
              return new RbacDataError("Failed to evaluate grant revocation for grant " + grant, e);
            });
  }

  private Set<PolicyGrant> getCandidates(
      SnowflakeObjectType grantObjectType, PolicyGrantPrivilege grantPrivilege) {
    return index
        .kv()
        .getOrDefault(grantObjectType.getAliasFor(), new ConcurrentHashMap<>())
        .getOrDefault(grantPrivilege, Set.of());
  }
}
